solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -1331,8 +1338,9 @@ public void process(WatchedEvent event) {
         return;
       }
 
-      if (!collectionWatches.watchedCollections().contains(coll)) {
-        // This collection is no longer interesting, stop watching.
+      StatefulCollectionWatch scw = collectionWatches.statefulWatchesByCollectionName.get(coll);
+      if (scw == null || scw.associatedWatcher != this) {
+        // Collection no longer interesting, or we have been replaced by a different watcher.
```

Comment:
The substance of the fix is in these three lines. Since StateWatcher instance do a deferred check against `collectionWatches` in order to determine whether to exit, it's possible that by the time a StateWatcher checks here, the watch it was created to serve has been removed, and a new watch inserted (and new StateWatcher created) before this check. In that case it's possible to accumulate persistent StateWatcher instances that all notify callbacks on the same event.

The fix is for the StateWatcher to check whether _it_ is the instance responsible for the current entry in `collectionWatches`, and exit if not. The rest of the changes are basically just to ensure that The new `StatefulCollectionWatch.associatedWatcher` field is never null once it's been inserted in the map.

Reply:
Maybe will do f2f session for it. 

`the watch it was created to serve has been removed, and a new watch inserted `

When/where we remove watch?

Reply:
The above 3 lines is where the `StateWatcher` is "removed", passively, by having the `process()` method return early without resetting itself. There are a number of places where `DocCollectionWatcher`s are started, including `waitForState()`, etc. Not actually sure which of these is the likely culprit, but the test definitely reproduces the problematic behavior we were observing, in a way that should not be possible to do (and is fixed by this change).